File: src/0904.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        unordered_map<int, int> fcount;
        int i = 0, j = 0;
        for(j; j < fruits.size(); j++) {
            fcount[fruits[j]]++;
            if(fcount.size() > 2) {
                if(--fcount[fruits[i]] == 0) {
                    fcount.erase(fruits[i]);
                }
                i++;
            }
        }
        return j-i;
    }
};

File: 0904.py
from collections import defaultdict

class Solution:
    def totalFruit(self, fruits: list[int]) -> int:
        fcount = defaultdict(int)
        i = 0
        for j in range(len(fruits)):
            fcount[fruits[j]] += 1
            if len(fcount) > 2:
                fcount[fruits[i]] -= 1
                if fcount[fruits[i]] == 0:
                    del fcount[fruits[i]]
                i += 1
        return j-i+1
